Contact field mapping

zpawn opened the pull request:
Story https://app.shortcut.com/deskpro/story/122828/hubspot-field-mapping

1. Implemented the creation of a structure using d'n'd in the admin side
2. Render contact custom properties on home page
3. Render Details contact page via BlocksBuilder
4. Сontact creation form is based on field mapping.
5. Update contact form based on field mapping
6. Сonfigured strict type-checking

<img width="682" alt="Screenshot 2024-10-25 at 14 28 01" src="https://github.com/user-attachments/assets/d1e51aab-16af-4f6f-9686-57337753ad4f">
<img width="267" alt="Screenshot 2024-10-25 at 14 33 37" src="https://github.com/user-attachments/assets/67b449b8-c4a5-4adc-b64f-171a4a0058a5">
<img width="267" alt="Screenshot 2024-10-25 at 14 34 01" src="https://github.com/user-attachments/assets/02840994-6d10-4bc0-b8d1-e54fbac92e60">

zpawn committed:
Implemented the creation of a structure using d'n'd in the admin side

HappyPaul55 commented on src/App.tsx:
```diff
@@ -56,34 +55,29 @@ function App() {
     }, [client, unlinkContactFn]);
 
-    if (!client || isLoading) {
+    if (!client || isLoading && !context) {
```
Why do we need this extra chcek. If it's loading, surely we just show the spinner?

zpawn replied:
You are right, no needed.

HappyPaul55 commented on src/services/hubspot/baseRequest.ts:
```diff
@@ -94,7 +73,15 @@ const baseRequest: Request = async (client, {
+      result = await res.json();
+    } catch (e) {
+      return {};
+    }
```
Do we need a console.warn here? It seems strange to expect a JSON and just return a blank object.

zpawn replied:
added error message
`console.warn("Failed to parse response as JSON. Returning empty result")`

HappyPaul55 commented on src/services/hubspot/baseRequest.ts:
```diff
@@ -94,7 +73,15 @@ const baseRequest: Request = async (client, {
+      return {};
+    }
+  
+    return result;
```
Minor: less complexity
```typescript
let result = {};
    try {
      result = await res.json();
    } catch (e) {
      // log warning
    }

    return result;
```

HappyPaul55 commented on src/utils/is.ts:
```diff
@@ -0,0 +1,14 @@
+// eslint-disable-next-line @typescript-eslint/no-explicit-any
+export const isPrimitive = (value: any): value is string | number => {
+    return isString(value) || isNumber(value);
+};
```
These can just be inlined, e.g.
```typescript
(typeof search === 'string' ? search.toLowerCase() : search).includes(searchInput.toLowerCase());
```

zpawn replied:
rewrote it like this
```js
`${search}`.toLowerCase().includes(searchInput.toLowerCase());
```

zpawn committed:
Render contact custom properties on home page

Story https://app.shortcut.com/deskpro/story/165035/hubspot-contact-field-mapping

zpawn committed:
Render Details contact page via BlocksBuilder

Story https://app.shortcut.com/deskpro/story/165035/hubspot-contact-field-mapping

zpawn committed:
Сontact creation form is based on field mapping.

Story https://app.shortcut.com/deskpro/story/165035/hubspot-contact-field-mapping

zpawn committed:
Update contact form based on field mapping

Story https://app.shortcut.com/deskpro/story/165035/hubspot-contact-field-mapping

zpawn committed:
DnD ui improvements

Story https://app.shortcut.com/deskpro/story/165035/hubspot-contact-field-mapping

HappyPaul55 commented on src/components/blocks/index.ts:
```diff
@@ -0,0 +1,32 @@
+    select: SelectBlock,
+    booleancheckbox: BooleanBlock,
+// eslint-disable-next-line @typescript-eslint/no-explicit-any
+} as Record<string, FC<any>>;
```
No need to type as Record here. Instead it should be `as const` as it doesn't change at all.

zpawn replied:
Fixed as `export const blocksMap = { ... } as const as BlocksMap`

HappyPaul55 commented on src/components/common/Builder/BlocksBuilder/RenderRow.tsx:
```diff
@@ -0,0 +1,55 @@
+          return null;
+        }
+      
+        if (!Component) {
+          // eslint-disable-next-line no-console
+          console.error(
+            "BlocksBuilder: can't find component for block type:",
+            blockType,
+          );
+        }
```
If `Component` isn't found, I assume we should return here as the below code will require one?

zpawn replied:
`Component` it's not necessary. If the `Component` isn't found, the value will be set to the default.

HappyPaul55 commented on src/components/common/Builder/StructureBuilder/AddItemDropdown.tsx:
```diff
@@ -0,0 +1,58 @@
+    const options = items.map((i) => getOption(i, (meta ?? {})[i]?.label ?? i));
+
+    const currentOptions = useMemo(() => {
+        return getFilteredOptions(options as Array<Option>, searchInput);
```
We shouldn't need to use `as` if our types are well defined.

zpawn replied:
fixed

HappyPaul55 commented on src/components/common/Builder/StructureBuilder/utils.ts:
```diff
@@ -0,0 +1,34 @@
+import { isPrimitive } from "../../../../utils";
+import type { DropdownItemType, DropdownValueType } from "@deskpro/deskpro-ui";
+
+const NO_FOUND = "No item(s) found";
```
`NOT_FOUND`

zpawn replied:
fixed

HappyPaul55 commented on src/components/common/Typography/DPNormalize.tsx:
```diff
@@ -0,0 +1,21 @@
+`;
+
+const DPNormalize: FC<Props> = ({ text }) => (
+  <Text dangerouslySetInnerHTML={{ __html: addBlankTargetToLinks(text) || "-" }} />
```
Where is the source of this `html` coming from, is it safe?

zpawn replied:
It can be a custom property with the type `html` https://developers.hubspot.com/docs/api/crm/properties

HappyPaul55 commented on src/components/fields/index.ts:
```diff
@@ -0,0 +1,30 @@
+    booleancheckbox: SelectField,
+    radio: SelectField,
+    checkbox: MultipleCheckboxesField,
+} as Record<PropertyMeta["fieldType"], FC<FieldProps>>;
```
`as const`

zpawn replied:
fixed

HappyPaul55 commented on src/constants.ts:
```diff
@@ -1 +1,33 @@
+export const TABS = {
+    HOME: 0,
+    VIEW: 1,
+};
```
`as const` or maybe an `enum` is better here.

zpawn replied:
fixed

HappyPaul55 commented on src/constants.ts:
```diff
@@ -1 +1,33 @@
+
+export const DndTypes = {
+    CONTACT: "contact",
+};
```
`as const` or maybe better as an enum.

zpawn replied:
fixed

HappyPaul55 commented on src/constants.ts:
```diff
@@ -1 +1,33 @@
+            ["lifecyclestage"],
+        ]
+    },
+};
```
`as const`

zpawn replied:
fixed

HappyPaul55 commented on src/hooks/useAsyncError.ts:
```diff
@@ -0,0 +1,16 @@
+import { useState, useCallback } from "react";
+
+const useAsyncError = () => {
+  // eslint-disable-next-line @typescript-eslint/no-unused-vars
+  const [_, setError] = useState();
+
+  const asyncErrorHandler = useCallback((e: Error) => {
+    setError(() => {
+      throw e;
+    });
+  }, [setError]);
+
+  return { asyncErrorHandler };
+};
```
Not sure i  understand the value of this function. Do we need to `useState`, why can't the calling code just throw the error?

zpawn replied:
Because `react-error-boundary` doesn't catch async errors. Therefore, we need to make them synchronous.

HappyPaul55 commented on src/types.ts:
```diff
@@ -62,12 +61,9 @@ export type DeskproUser = {
+    app: object;
+    currentAgent: object,
+    env: object,
     user: DeskproUser,
```
This should be strongly typed, we know the format of this so we should declare it ahead of time.

zpawn replied:
deleted `app`, `env` and `currentAgent ` because we aren't using them.

HappyPaul55 reviewed (changes requested):
Please can you update the ESLint rule to use `plugin:@typescript-eslint/strict-type-checked`. This PR is so big I don't feel confident in merging it without the thumbs up from the linter as well.

zpawn committed:
Fixed comments

zpawn committed:
Сonfigured strict type-checking

HappyPaul55 commented on jest.setup.tsx:
```diff
@@ -0,0 +1,65 @@
+global.React = React;
+// eslint-disable-next-line @typescript-eslint/ban-ts-comment
+//@ts-ignore
+global.ResizeObserver = ResizeObserver;
```
What errors does it throw and is there a workaround that doesn't require ts/es lint ignores?

zpawn replied:
fixed

HappyPaul55 commented on src/components/Deal/DealInfo.tsx:
```diff
@@ -1,25 +1,23 @@
 const DealInfo: FC<Props> = ({ deal, pipeline, accountInfo, owner, dealTypes }) => {
-    const stage = pipeline.stages.find(({ id }) => id === deal.dealstage) || {};
-    const dealType = dealTypes?.options.find(({ value }) => value === deal.dealtype);
+    const stage = pipeline.stages?.find(({ id }) => id === deal.dealstage) || {} as PipelineStage;
```
Avoid `as`, `const stage: PipelineStage = ...;` is better as it ensures the object is satisfies properly

zpawn replied:
fixed

HappyPaul55 commented on src/components/blocks/index.ts:
```diff
@@ -0,0 +1,31 @@
+    checkbox: MultipleCheckboxesBlock,
+    select: SelectBlock,
+    booleancheckbox: BooleanBlock,
+} as const as BlocksMap;
```
`as const` should be all that's needed. Retyping it to `BlocksMap` makes it wider again.

zpawn replied:
fixed

HappyPaul55 commented on src/components/common/Builder/BlocksBuilder/types.ts:
```diff
@@ -0,0 +1,23 @@
+import type { PropertyMeta } from "../../../../services/hubspot/types";
+import type { Layout } from "../types";
+
+export type BlocksMap = Record<PropertyMeta["fieldType"], FC<BlockProps>>;
```
Blocks map can just be typeof https://github.com/DeskproApps/hubspot/pull/45/files#diff-4ecc753abf4cfb0dadad462fb5866d83ad43a919bf30f69b6ad55cc892dc8613R10

zpawn replied:
Unfortunately, we can't. We don't know in advance what the `blocksMap` might be.

HappyPaul55 replied:
Is BlocksMap not related to https://github.com/DeskproApps/hubspot/pull/45/files/21bb0840cc872b28e057a33559b2a3bfc147cdd2#diff-4ecc753abf4cfb0dadad462fb5866d83ad43a919bf30f69b6ad55cc892dc8613?

zpawn replied:
The map is passed from the outside BlocksBuilder, and it can be anything. And inside the block receives the value and metadata.

HappyPaul55 commented on src/components/common/Builder/StructureBuilder/StructureBuilder.tsx:
```diff
@@ -0,0 +1,101 @@
+            return;
+        }
+    
+        let newRows = [...rows];
```
const

zpawn replied:
Two lines below, we save the new value

HappyPaul55 commented on src/components/fields/index.ts:
```diff
@@ -0,0 +1,28 @@
+    booleancheckbox: SelectField,
+    radio: SelectField,
+    checkbox: MultipleCheckboxesField,
+} as const as FieldsMap;
```
no need for `as FieldsMap` as that widens the scope

zpawn replied:
fixed

HappyPaul55 commented on src/constants.ts:
```diff
@@ -1 +1,35 @@
+            ["phone"],
+            ["hubspot_owner_id"],
+            ["lifecyclestage"],
+        ] as Layout,
```
No need for `as`

zpawn replied:
Unfortunately, they are needed here. In the app, we are trying to get this structure through `JSON.parse`, and TypeScript can't match it with this type.

HappyPaul55 commented on src/constants.ts:
```diff
@@ -1 +1,35 @@
+            ["phone"],
+            ["hubspot_owner_id"],
+            ["lifecyclestage"],
+        ] as Layout,
```
No need for `as`

zpawn replied:
Unfortunately, they are needed here. In the app, we are trying to get this structure through `JSON.parse`, and TypeScript can't match it with this type.

HappyPaul55 commented on src/hooks/AppContext.tsx:
```diff
@@ -0,0 +1,33 @@
+  const [settings, setSettings] = useState<Settings|null>(null);
+  const [isLoading, setIsLoading] = useState<boolean>(true);
+
+  useEffect(() => setIsLoading(!settings), [settings]);
```
No need for state here, use `const isLoading = !settings;`

zpawn replied:
fixed

HappyPaul55 commented on src/pages/HomePage/hooks.ts:
```diff
@@ -139,16 +157,24 @@ const useLoadHomeDeps = (contactId: Contact["id"]|null) => {
     return {
-        isLoading: [contact].every(({ isLoading }) => Boolean(isLoading)),
+        isLoading: [linkedContactIds, contact, propertiesMeta].some(({ isLoading }) => isLoading),
         contact: get(contact, ["data", "properties"], {}) as Contact["properties"],
```
This type is incorrect, if a blank object is used `{}` then that doesn't match the type that `Contact["properties"]` should be as it doesn't include some fields like `email`, `phone` [and more](https://github.com/DeskproApps/hubspot/blob/f87d03a35a530a854a1ab6bc693203eb604ddd13/src/services/hubspot/types.ts#L44)

zpawn replied:
fixed

HappyPaul55 commented on src/pages/HomePage/hooks.ts:
```diff
@@ -139,16 +157,24 @@ const useLoadHomeDeps = (contactId: Contact["id"]|null) => {
         companies: filterEntities(companies) as Array<Company["properties"]>,
         deals: deals.data || [],
-        dealPipelines: dealPipelinesData,
+        dealPipelines: dealPipelinesData as Record<Pipeline["id"], Pipeline>,
```
If the type is correct, it shouldn't need an `as` typecast.

zpawn replied:
fixed

HappyPaul55 commented on src/services/hubspot/baseRequest.ts:
```diff
@@ -1,17 +1,17 @@
     method: ApiRequestMethod,
     code: number,
-    json?: HubSpotError,
+    json?: HubSpotError|null,
```
`|undefined` is probably better as it reduces the number of types you ultimately have to deal with.

zpawn replied:
fixed

HappyPaul55 commented on src/utils/is.ts:
```diff
@@ -0,0 +1,32 @@
+// eslint-disable-next-line @typescript-eslint/no-explicit-any
+export const isString = (value: any): value is string => {
+    return typeof value === "string";
+};
```
I personally don't like these, it adds no value compared to just doing `typeof value === "string"` each time, or argubably more confusing what it's actually doing is hidden away in this "helper" function.

zpawn replied:
removed

HappyPaul55 commented on src/utils/is.ts:
```diff
@@ -0,0 +1,32 @@
+// eslint-disable-next-line @typescript-eslint/no-explicit-any
+export const isNumber = (value: any): value is number => {
+    return typeof value === "number" && !isNaN(value);
+};
```
For the same reason above, I'd avoid doing this as this can cause confusion. `NaN` is a `number` some of the time and by hiding that logic in this "helper" it makes it more guess work what is going on.

zpawn replied:
removed

HappyPaul55 commented on src/utils/is.ts:
```diff
@@ -0,0 +1,32 @@
+};
+
+// eslint-disable-next-line @typescript-eslint/no-explicit-any
+export const isPrimitive = (value: any): value is string | number => {
```
Not sure that just numbers and strings a primitives, are `null` and `boolean` ones too? It's probably best to no abstract this to helper functions just write what you need at the tim.

zpawn replied:
removed

HappyPaul55 commented on src/utils/is.ts:
```diff
@@ -0,0 +1,32 @@
+};
+
+export const isEmptyObject = (obj: object): boolean => {
+    return obj == null || (Object.keys(obj).length === 0 && obj.constructor === Object);
```
With typescript this seems overkill. `Object.values(input).length === 0` is enough (let typescript handle the types and let code handle the data.

zpawn replied:
fixed

HappyPaul55 commented on src/utils/tryToLinkAutomatically.ts:
```diff
@@ -0,0 +1,45 @@
+    try {
+        const contactIds = await getEntityContactList(client, dpUser.id);
+
+        if (contactIds?.length === 1) {
```
Can you add comments on why these checks exist? I'm not too sure what's going on here.

zpawn replied:
done

HappyPaul55 replied:
![image](https://github.com/user-attachments/assets/74a4538d-f155-4425-9946-cad150c93f06)
I see no comments?

zpawn committed:
Fixed comments
